packages/dataDisplay/useSortableTable/src/useSortableTable.tsx

```diff
@@ -0,0 +1,109 @@
+import { useCallback, useEffect, useMemo, useState } from 'react';
+
+type SortableTableDataType<V extends Object> = Map<keyof V, Array<V[keyof V]>>;
+
+const _initalizeConvertedData = <V extends Object,>(data: V[], keys: (keyof V)[]) => {
+    const initTableData: SortableTableDataType<V> = new Map();
+
+    keys.forEach((key) => {
+        initTableData.set(key, []);
+    });
+
+    data.forEach((rowData) => {
+        (Object.keys(rowData) as (keyof V)[]).forEach((key) => {
+            const targetMapArray = initTableData.get(key);
+
+            if (targetMapArray === undefined) {
+                throw new Error('Can not find key in data');
+            }
+
+            targetMapArray.push(rowData[key]);
+        });
+    });
+    return initTableData;
+}
+
+const useSortableTable = <V extends Object>(data: V[]) => {
+    type K = keyof V;
+
+    const [convertedData, setConvertedData] = useState<SortableTableDataType<V> | undefined>(undefined);
+    const [sortedData, setSortedData] = useState<V[]>(() => data);
+
+    useEffect(() => {
+        if (data.length === 0) {
+            return;
+        }
+        setSortedData(data);
+        setConvertedData(_initalizeConvertedData(data, keys));
+    }, [data]);
+
+    const keys = useMemo(() => {
+        const objectKeys = data.reduce((acc, currentValue) => {
+            const currentValueKeys = Object.keys(currentValue) as K[];
+
+            const newObjectKeys = currentValueKeys.filter((key) => !acc.includes(key));
+
+            return [...acc, ...newObjectKeys];
+        }, [] as K[]);
+
+        return objectKeys;
+    }, [data]);
+
+    const sort = useCallback((key: K, compareFn: (a: any, b: any) => number) => {
+        if (keys.length === 0) {
+            return;
+        }
+
+        if (convertedData === undefined) {
+            throw new Error('tableData is not initialized');
+        }
+
+        const targetColumn = convertedData.get(key);
+
+        if (targetColumn === undefined) {
+            throw new Error('Can not find key in data');
+        }
+
+        const sortedIndex = targetColumn
+            .map((v, i) => ({ v, i }))
+            .sort((a, b) => compareFn(a.v, b.v))
+            .map((sorted) => sorted.i);
+
+        const newSortedData = sortedIndex.map((index) =>
+            keys.map((key) => {
+                const value = convertedData.get(key);
+
+                if (!value) {
+                    throw new Error('Can not find key in data');
+                }
+
+                return { [key]: value[index] };
+            }).reduce((acc, current) => {
+                let ret = Object.assign({}, acc);
+                ret = Object.assign(ret, current);
+
+                return ret;
+            }, {} as V)
```

Comment:
```suggestion
            }).reduce((acc, current) => {
                return {...acc, ...current};
            }, {} as V)
```
이렇게 좀더 축약할 수도 있겠네요!

그리고 중요하진 않지만 이런 교환법칙 성립하는 바이너리 오퍼레이터를 쓸때는 lhs(left hand side), rhs(right hand side)라는 표현을 많이 씁니다 ㅎ

```suggestion
            }).reduce((lhs, rhs) => {
                return {...lhs, ...rhs};
            }, {} as V)
```

Reply:
> 그리고 중요하진 않지만 이런 교환법칙 성립하는 바이너리 오퍼레이터를 쓸때는 lhs(left hand side), rhs(right hand side)라는 표현을 많이 씁니다 ㅎ

아하 종종 다른 코드 참고할 때 많이 보던 변수명이었습니다! 추천해주신 코드로 수정하도록 하겠습니다!